base_layer/core/src/base_node/states/horizon_state_sync/validators/headers.rs

```diff
@@ -0,0 +1,250 @@
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    blocks::{BlockHeader, BlockHeaderValidationError},
+    chain_storage::{BlockchainBackend, BlockchainDatabase},
+    consensus::ConsensusManager,
+    proof_of_work::{get_median_timestamp, get_target_difficulty, Difficulty, PowError},
+    validation::{StatelessValidation, ValidationError},
+};
+use log::*;
+use tari_crypto::tari_utilities::{epoch_time::EpochTime, hex::Hex, Hashable};
+
+const LOG_TARGET: &str = "c::bn::states::horizon_state_sync::headers";
+
+pub struct HorizonHeaderValidator<B> {
+    rules: ConsensusManager,
+    db: BlockchainDatabase<B>,
+}
+
+impl<B: BlockchainBackend> HorizonHeaderValidator<B> {
+    pub fn new(db: BlockchainDatabase<B>, rules: ConsensusManager) -> Self {
+        Self { db, rules }
+    }
+}
+
+impl<B: BlockchainBackend> StatelessValidation<BlockHeader> for HorizonHeaderValidator<B> {
+    fn validate(&self, header: &BlockHeader) -> Result<(), ValidationError> {
+        let header_id = format!("header #{} ({})", header.height, header.hash().to_hex());
+        let tip_header = self
+            .db
+            .fetch_tip_header()
+            .map_err(|e| ValidationError::CustomError(e.to_string()))?;
+        self.check_median_timestamp(header, &tip_header)?;
+        trace!(
+            target: LOG_TARGET,
+            "BlockHeader validation: Median timestamp is ok for {} ",
+            &header_id
+        );
+        self.check_achieved_and_target_difficulty(header, &tip_header)?;
+        trace!(
+            target: LOG_TARGET,
+            "BlockHeader validation: Achieved difficulty is ok for {} ",
+            &header_id
+        );
+        debug!(
+            target: LOG_TARGET,
+            "Block header validation: BlockHeader is VALID for {}", &header_id
+        );
+
+        Ok(())
+    }
+}
+
+impl<B: BlockchainBackend> HorizonHeaderValidator<B> {
+    pub fn is_genesis(&self, block_header: &BlockHeader) -> bool {
+        block_header.height == 0 && self.rules.get_genesis_block_hash() == block_header.hash()
+    }
+
+    /// Calculates the achieved and target difficulties at the specified height and compares them.
+    pub fn check_achieved_and_target_difficulty(
+        &self,
+        block_header: &BlockHeader,
+        tip_header: &BlockHeader,
+    ) -> Result<(), ValidationError>
+    {
+        let pow_algo = block_header.pow.pow_algo;
+        let target = if self.is_genesis(block_header) {
+            Difficulty::from(1)
+        } else {
+            let target_difficulties = self.fetch_target_difficulties(block_header, tip_header)?;
```

Comment:
Looking at this code, this will fail, if your header.height = 1, and the tip header.height = 2.